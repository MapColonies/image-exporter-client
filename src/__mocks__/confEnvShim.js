if (!window._env_) {
  window._env_ = (function (undefined) {
    return {
      LANGUAGE: 'en',
      MAP_SERVER: 'MAP_SERVER',
      PUBLISH_POINT: 'PUBLISH_POINT',
      CHANNEL: 1002,
      VERSION: 1,
      REQUEST: 'REQUEST',
      SERVICE_PROTOCOL: 'SERVICE_PROTOCOL',
      SERVICE_NAME: 'SERVICE_NAME',
      ACTIVE_LAYER: 'ACTIVE_LAYER',
      BBOX: {
        limit: 20000,
      },
      EXPORT: {
        defaultZoomLevel: 18,
        maxZoomLevel: 22
      },
      MAP: {
        center: [35.14, 31.39],
        zoom: 8,
      },
      ACTIVE_LAYER_PROPERTIES: {
        urlPattern : 'MAP_SERVER/service',
        exportUrlPattern: 'MAP_SERVER/service?REQUEST=REQUEST&SERVICE=SERVICE&LAYERS=LAYER',
        urlPatternParams: {
          service: 'SERVICE',
          layers: 'LAYER',
          tiled: 'TILED',
          matrixSet: 'MATRIXSET',
          style: 'STYLE',
          projection: 'PROJECTION'
        }
      },
    };
  })(void 0);
}
