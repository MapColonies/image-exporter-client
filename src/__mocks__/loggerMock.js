export default {
  log: () => {},
  info: () => {},
  error: () => {},
  debug: () => {},
};
