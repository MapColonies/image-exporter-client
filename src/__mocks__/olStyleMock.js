export class Fill {
  constructor() { }
}

export class Stroke {
  constructor() { }
}

export class Circle {
  constructor() { }
}

export class Style {
  constructor() { }
}
