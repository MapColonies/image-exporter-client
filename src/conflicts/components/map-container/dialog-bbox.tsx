import React, { useState } from 'react';
import * as  turf from '@turf/helpers';
import bbox from '@turf/bbox';
import bboxPolygon from '@turf/bbox-polygon';
import distance  from '@turf/distance/dist/js'; //TODO: make a consumption "REGULAR"
import { Polygon } from 'geojson';
import { useFormik } from 'formik';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import { 
  Dialog, 
  DialogTitle, 
  DialogContent, 
  TextField, 
  Button} from '@map-colonies/react-core';
import { Box } from '@map-colonies/react-components';
import EXPORTER_CONFIG from '../../../common/config';

  const useStyle = makeStyles((theme: Theme) =>
  createStyles({
    bbox:{
      overflow: 'hidden',
      position: 'relative',
      border: 'solid 1px', 
      width: '60px', 
      height: '56px',
    },

    bboxLeftBottomCorner: {
      "&::before": {
        margin: '-1em',
        borderRadius: '50%',
        position: 'absolute',
        padding: '1em',
        boxShadow: '0 0 7px #b53',
        background: '#95a',
        content: "''",
        bottom: 0   
      }
    },
    bboxRightTopCorner: {
      "&::before": {
        margin: '-1em',
        borderRadius: '50%',
        position: 'absolute',
        padding: '1em',
        boxShadow: '0 0 7px #b53',
        background: '#95a',
        content: "''",
        top: 0,
        right: 0   
      }
    },
    spacer: {
      marginRight: '16px'
    },
    errorContainer: {
      display: 'flex',
      alignItems: 'center',
      marginRight: 'auto',
      color: theme.palette.error.main,
    }

  })
);

interface BBoxCorners {
  bottomLeftLat: number;
  bottomLeftLon: number;
  topRightLat: number;
  topRightLon: number;
}

interface BBoxCornersError {
  latDistance: string;
  lonDistance: string;
}

const validate = (values: BBoxCorners): BBoxCornersError => {
  const errors: BBoxCornersError = {latDistance: '', lonDistance: ''};

  try{
    turf.lineString([
      [values.bottomLeftLon, values.bottomLeftLat],
      [values.topRightLon, values.topRightLat], 
    ]);

    const yDistance = distance(
      [values.bottomLeftLon, values.bottomLeftLat],
      [values.bottomLeftLon, values.topRightLat]);
    
    const xDistance = distance(
      [values.bottomLeftLon, values.bottomLeftLat],
      [values.topRightLon, values.bottomLeftLat]);
  
    if (xDistance > EXPORTER_CONFIG.BOUNDARIES.MAX_X_KM){
      errors.latDistance = 'X distance is exceeded the limit';
    }
    if (yDistance > EXPORTER_CONFIG.BOUNDARIES.MAX_Y_KM){
      errors.lonDistance = 'Y distance is exceeded the limit';
    }
  }
  catch(err){
    errors.latDistance = 'Not valid coordinates';
  }

  return errors;
};


interface DialogBBoxProps {
  isOpen: boolean;
  onSetOpen: (open: boolean) => void;
  onPolygonUpdate: (polygon: Polygon) => void;
}

export const DialogBBox: React.FC<DialogBBoxProps> = (
  props
) => {
  const { isOpen, onSetOpen, onPolygonUpdate } = props;
  const classes = useStyle();
  const formik = useFormik({
    initialValues: {
      bottomLeftLat: 0,
      bottomLeftLon: 0,
      topRightLat: 0,
      topRightLon: 0,
    },
    onSubmit: values => {
      const err = validate(values);
      if(!err.latDistance && !err.lonDistance){
        const line = turf.lineString([
          [values.bottomLeftLon, values.bottomLeftLat],
          [values.topRightLon, values.topRightLat], 
        ]);
        const polygon = bboxPolygon(bbox(line));
        console.log('polygon', polygon.geometry);
  
        onPolygonUpdate(polygon.geometry);
        handleClose(false);
      }
      else{
        setFormErrors(err);
      }

    },
  });

  const [formErrors, setFormErrors] = useState({latDistance: '', lonDistance: ''});
  
  const handleClose = (isOpened:boolean):void => {
    onSetOpen(isOpened);
  }
  return (
      <Dialog open={isOpen}>
        <DialogTitle>Define bbox coordinates</DialogTitle>
        <DialogContent>
        <form onSubmit={formik.handleSubmit}>
          <Box style={{display: 'flex', marginBottom: '16px'}}>
            <TextField 
              label="Top Right Lat" 
              id="topRightLat"
              name="topRightLat"
              type="number"
              onChange={formik.handleChange}
              value={formik.values.topRightLat}
              className={classes.spacer}
            />
            <TextField 
              label="Top Right Lon" 
              id="topRightLon"
              name="topRightLon"
              type="number"
              onChange={formik.handleChange}
              value={formik.values.topRightLon}
              className={classes.spacer}
            />
            <div className={`${classes.bbox} ${classes.bboxRightTopCorner}`}></div>
          </Box>
          <Box style={{display: 'flex'}}>
          <TextField 
              label="Bottom Left Lat" 
              id="bottomLeftLat"
              name="bottomLeftLat"
              type="number"
              onChange={formik.handleChange}
              value={formik.values.bottomLeftLat}
              className={classes.spacer}
            />
            <TextField 
              label="Bottom Left Lon" 
              id="bottomLeftLon"
              name="bottomLeftLon"
              type="number"
              onChange={formik.handleChange}
              value={formik.values.bottomLeftLon}
              className={classes.spacer}
            />
            <div className={`${classes.bbox} ${classes.bboxLeftBottomCorner}`}></div>
          </Box>
          <Box style={{ display: 'flex', justifyContent: 'flex-end', marginTop: '16px', gap: '16px' }}>
            {
              (!!formErrors.latDistance || !!formErrors.lonDistance) ? 
              <div className={classes.errorContainer}>
                {`Error: ${formErrors.latDistance} ${formErrors.lonDistance}`}
              </div> : 
              null
            }
            <Button type="button" onClick={()=>{handleClose(false);}}>Cancel</Button>
            <Button raised type="submit">Ok</Button>
          </Box>
        </form>
        </DialogContent>
      </Dialog>

  );
}
