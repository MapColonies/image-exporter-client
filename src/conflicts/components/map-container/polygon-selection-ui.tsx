import React from 'react';
import { Polygon } from 'geojson';
import { Menu, 
  MenuItem,
  Button,
  Tooltip,
  Dialog, 
  DialogTitle, 
  DialogContent, 
  DialogActions, 
  DialogButton } from '@map-colonies/react-core';
import '@map-colonies/react-core/dist/button/styles';
import '@map-colonies/react-core/dist/tooltip/styles';
import '@map-colonies/react-core/dist/menu/styles';

import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

import { Box, DrawType } from '@map-colonies/react-components';

const WIDTH_SPACING_FACTOR = 18;
const useStyle = makeStyles((theme: Theme) =>
  createStyles({
    drawingButton: {
      width: theme.spacing(WIDTH_SPACING_FACTOR),
    },
    fullWidth: {
      width: '100%',
      marginTop: '36px',
    },
  })
);

export interface PolygonSelectionUiProps {
  isSelectionEnabled: boolean;
  onStartDraw: (type: DrawType) => void;
  onCancelDraw: () => void;
  onReset: () => void;
  onPolygonUpdate: (polygon: Polygon) => void;
}

export const PolygonSelectionUi: React.FC<PolygonSelectionUiProps> = (
  props
) => {
  const classes = useStyle();
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  const { isSelectionEnabled, onCancelDraw, onStartDraw, onReset, onPolygonUpdate } = props;

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>): void => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = (): void => {
    setAnchorEl(null);
  };

  const [open, setOpen] = React.useState(false);

  if (isSelectionEnabled) {
    return (
      <Tooltip content="Cancel the ongoing draw" align={'bottomLeft'}>
        <Button className={classes.drawingButton} raised onClick={onCancelDraw}>
          Cancel Draw
        </Button>
      </Tooltip>
    );
  } else {
    return (
      <Box position="relative">
        <Tooltip
          content="draw an Area of interest to limit the search"
          align={'bottomLeft'}
        >
          <Button
            className={classes.drawingButton}
            raised
            onClick={handleClick}
          >
            Draw AOI
          </Button>
        </Tooltip>
        <Dialog
          open={open}
          onClose={evt => {
            const SAMPLE_POLYGON: Polygon={
              type:"Polygon",
              coordinates:[[
                  [34.70371442859832,32.022954156119006],
                  [35.146463443653914,32.022954156119006],
                  [35.146463443653914,32.23608887564354],
                  [34.70371442859832,32.23608887564354],
                  [34.70371442859832,32.022954156119006]
              ]]
            };
          
            console.log(evt.detail.action);
            onPolygonUpdate(SAMPLE_POLYGON);
            setOpen(false);
          }}
          onClosed={evt => console.log(evt.detail.action)}
        >
          <DialogTitle>Define bbox coordinates</DialogTitle>
          <DialogContent>This is a standard dialog.</DialogContent>
          <DialogActions>
            <DialogButton action="close">Cancel</DialogButton>
            <DialogButton raised action="accept" isDefaultAction>
              Ok
            </DialogButton>
          </DialogActions>
        </Dialog>
        <Menu
          className={classes.fullWidth}
          open={Boolean(anchorEl)}
          onClose={handleClose}
        >
          {/* <MenuItem
            onClick={(): void => {
              onStartDraw(DrawType.POLYGON);
              handleClose();
            }}
          >
            Polygon
          </MenuItem> */}
          <MenuItem
            onClick={(): void => {
              onStartDraw(DrawType.BOX);
              handleClose();
            }}
          >
            Box
          </MenuItem>
          <MenuItem
            onClick={(): void => {
              // onStartDraw(DrawType.BOX);
              setOpen(true);
              handleClose();
            }}
          >
            Box by coordinates
          </MenuItem>
          <MenuItem
            onClick={(): void => {
              onReset();
              handleClose();
            }}
          >
            Clear
          </MenuItem>
        </Menu>
      </Box>
    );
  }
};
