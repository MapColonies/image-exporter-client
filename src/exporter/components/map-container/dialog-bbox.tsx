import React, { useState } from 'react';
import * as  turf from '@turf/helpers';
import bbox from '@turf/bbox';
import bboxPolygon from '@turf/bbox-polygon';
import { Polygon } from 'geojson';
import { useFormik } from 'formik';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  TextField,
  Button
} from '@map-colonies/react-core';
import { Box } from '@map-colonies/react-components';
import { FormattedMessage, useIntl, IntlShape } from 'react-intl';
import { BBoxCorner, Corner } from '../bbox/bbox-corner-indicator';
import { BBoxAreaLimit, isBBoxWithinLimit } from '../../../common/helpers/bbox-area';
import EXPORTER_CONFIG from '../../../common/config';

const useStyle = makeStyles((theme: Theme) =>
  createStyles({
    spacer: {
      marginRight: '16px'
    },
    errorContainer: {
      display: 'flex',
      alignItems: 'center',
      marginRight: 'auto',
      color: theme.palette.error.main,
    }

  })
);

interface BBoxCorners {
  bottomLeftLat: number;
  bottomLeftLon: number;
  topRightLat: number;
  topRightLon: number;
}

interface BBoxError {
  bboxArea: string;
}

const validate = (values: BBoxCorners, intl: IntlShape): BBoxError => {
  const errors: BBoxError = { bboxArea: '' };

  try {
    const line = turf.lineString([
      [values.bottomLeftLon, values.bottomLeftLat],
      [values.topRightLon, values.topRightLat],
    ]);
    const polygon = bboxPolygon(bbox(line));
    const isWithinLimit = isBBoxWithinLimit(polygon.geometry);
    if (isWithinLimit !== BBoxAreaLimit.OK) {
      const messageFormat = isWithinLimit === BBoxAreaLimit.TOO_BIG ? 'custom-bbox.form-error.area.large.text' : 'custom-bbox.form-error.area.small.text';
      errors.bboxArea = intl.formatMessage({ id: messageFormat });
    }
  } catch (err) {
    errors.bboxArea = intl.formatMessage({ id: 'custom-bbox.form-error.area.invalid.text' });
  }

  return errors;
};


interface DialogBBoxProps {
  isOpen: boolean;
  onSetOpen: (open: boolean) => void;
  onPolygonUpdate: (polygon: Polygon) => void;
}

export const DialogBBox: React.FC<DialogBBoxProps> = (
  props
) => {
  const { isOpen, onSetOpen, onPolygonUpdate } = props;
  const classes = useStyle();
  const intl = useIntl();
  const formik = useFormik({
    initialValues: {
      bottomLeftLat: 0,
      bottomLeftLon: 0,
      topRightLat: 0.00001,
      topRightLon: 0.00001,
    },
    onSubmit: values => {
      const err = validate(values, intl);

      if (!err.bboxArea) {
        const line = turf.lineString([
          [values.bottomLeftLon, values.bottomLeftLat],
          [values.topRightLon, values.topRightLat],
        ]);
        const polygon = bboxPolygon(bbox(line));

        onPolygonUpdate(polygon.geometry);
        handleClose(false);
        setFormErrors({
          bboxArea: ''
        });
      } else {
        setFormErrors(err);
      }

    },
  });

  const [formErrors, setFormErrors] = useState({ bboxArea: '' });

  const handleClose = (isOpened: boolean): void => {
    onSetOpen(isOpened);
  };

  // eslint-disable-next-line
  const checkValidLat = (e: React.ChangeEvent<any>) => {
    const maxMinLatValue = EXPORTER_CONFIG.MAP.MAX_MIN_LAT_VALUE;
    const isValid: boolean = isValidRegexPattern(e, maxMinLatValue);
    return isValid? formik.handleChange(e) : false;
  };

  // eslint-disable-next-line
  const checkValidLong = (e: React.ChangeEvent<any>) => {
    const maxMinLongValue = EXPORTER_CONFIG.MAP.MAX_MIN_LONG_VALUE;
    const isValid: boolean = isValidRegexPattern(e, maxMinLongValue);
    return isValid? formik.handleChange(e) : false;
  };

  const isValidRegexPattern = (e: React.ChangeEvent<any>, maxMinValue: number): boolean => {
    // eslint-disable-next-line
    const valueString: string = e.target.value as string;
    const splitChar = EXPORTER_CONFIG.EXPORT.SPLIT_CHARACTER;
    const maxFractionDigits = EXPORTER_CONFIG.EXPORT.MAX_FRACTION_DIGITS;
    const regexStr = `^([0-9])+\\${splitChar}?(\\b[0-9]{0,${maxFractionDigits}})*$`;
    const regex = new RegExp(regexStr);
    const value = Number(valueString);
    if (!valueString) {
      return false;
    }
    if (value < -maxMinValue || value > maxMinValue) {
      return false;
    }
    return regex.test(valueString);
  };

  return (
    <Dialog open={ isOpen } preventOutsideDismiss={ true }>
      <DialogTitle>
        <FormattedMessage id="custom-bbox.dialog.title" />
      </DialogTitle>
      <DialogContent>
        <form onSubmit={ formik.handleSubmit }>
          <Box style={ { display: 'flex', marginBottom: '16px' } }>
            <TextField
              label={ intl.formatMessage({ id: 'custom-bbox.dialog-field.top_right_lat.label' }) }
              id="topRightLat" 
              name="topRightLat"
              type="text"
              onChange={ checkValidLat }
              value={ formik.values.topRightLat }
              className={ classes.spacer }
            />
            <TextField
              label={ intl.formatMessage({ id: 'custom-bbox.dialog-field.top_right_lon.label' }) }
              id="topRightLon"
              name="topRightLon"
              type="text"
              onChange={ checkValidLong }
              value={ formik.values.topRightLon }
              className={ classes.spacer }
            />
            <BBoxCorner corner={ Corner.TOP_RIGHT } />
          </Box>
          <Box style={ { display: 'flex' } }>
            <TextField
              label={ intl.formatMessage({ id: 'custom-bbox.dialog-field.bottom_left_lat.label' }) }
              id="bottomLeftLat"
              name="bottomLeftLat"
              type="text"
              onChange={ checkValidLat }
              value={ formik.values.bottomLeftLat }
              className={ classes.spacer }
            />
            <TextField
              label={ intl.formatMessage({ id: 'custom-bbox.dialog-field.bottom_left_lon.label' }) }
              id="bottomLeftLon"
              name="bottomLeftLon"
              type="text"
              onChange={ checkValidLong }
              value={ formik.values.bottomLeftLon }
              className={ classes.spacer }
            />
            <BBoxCorner corner={ Corner.BOTTOM_LEFT } />
          </Box>
          <Box style={ { display: 'flex', justifyContent: 'flex-end', marginTop: '16px', gap: '16px' } }>
            {
              formErrors.bboxArea ?
                <div id="errorContainer" className={ classes.errorContainer }>
                  { `${ intl.formatMessage({ id: 'general.error.label' }) }: ${ formErrors.bboxArea }` }
                </div> :
                null
            }
            <Button type="button" onClick={ (): void => {
              handleClose(false);
            } }>
              <FormattedMessage id="general.cancel-btn.text" />
            </Button>
            <Button raised type="submit">
              <FormattedMessage id="general.ok-btn.text" />
            </Button>
          </Box>
        </form>
      </DialogContent>
    </Dialog>

  );
};
