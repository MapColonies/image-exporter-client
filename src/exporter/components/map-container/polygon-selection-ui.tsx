import React from 'react';
import { Polygon } from 'geojson';
import { Menu, 
  MenuItem,
  Button,
  Tooltip } from '@map-colonies/react-core';
import '@map-colonies/react-core/dist/button/styles';
import '@map-colonies/react-core/dist/tooltip/styles';
import '@map-colonies/react-core/dist/menu/styles';

import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

import { Box, DrawType } from '@map-colonies/react-components';
import { useIntl, FormattedMessage } from 'react-intl';
import { DialogBBox } from './dialog-bbox';

const WIDTH_SPACING_FACTOR = 18;
const useStyle = makeStyles((theme: Theme) =>
  createStyles({
    drawingButton: {
      width: theme.spacing(WIDTH_SPACING_FACTOR),
    },
    fullWidth: {
      width: '100%',
      marginTop: '36px',
    },
  })
);

export interface PolygonSelectionUiProps {
  isSelectionEnabled: boolean;
  isDrawDisabled: boolean;
  onStartDraw: (type: DrawType) => void;
  onCancelDraw: () => void;
  onReset: () => void;
  onPolygonUpdate: (polygon: Polygon) => void;
}

export const PolygonSelectionUi: React.FC<PolygonSelectionUiProps> = (
  props
) => {
  const classes = useStyle();
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  const intl = useIntl();
  const { isSelectionEnabled, onCancelDraw, onStartDraw, onReset, onPolygonUpdate } = props;

  const handleClick = (event: React.MouseEvent<HTMLButtonElement>): void => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = (): void => {
    setAnchorEl(null);
  };

  const [open, setOpen] = React.useState(false);

  if (isSelectionEnabled) {
    return (
      <Tooltip 
        content={intl.formatMessage({id: 'polygon-selection.cancel-btn.tooltip'})} 
        align={'bottomLeft'}
      >
        <Button className={classes.drawingButton} raised onClick={onCancelDraw}>
          <FormattedMessage id="polygon-selection.cancel-btn.text"/>
        </Button>
      </Tooltip>
    );
  } else {
    return (
      <Box position="relative">
        <Tooltip
          content={intl.formatMessage({id: 'polygon-selection.draw-btn.tooltip'})}
          align={'bottomLeft'}
        >
          <Button
            className={classes.drawingButton}
            raised
            disabled={props.isDrawDisabled}
            onClick={handleClick}
          >
            <FormattedMessage id="polygon-selection.draw-btn.text"/>
          </Button>
        </Tooltip>
        <DialogBBox 
          isOpen={open}
          onSetOpen={setOpen}
          onPolygonUpdate={onPolygonUpdate}>
        </DialogBBox>
        <Menu
          className={classes.fullWidth}
          open={Boolean(anchorEl)}
          onClose={handleClose}
        >
          <MenuItem
            onClick={(): void => {
              onStartDraw(DrawType.BOX);
              handleClose();
            }}
          >
            <FormattedMessage id="polygon-selection.box-menu_option.text"/>
          </MenuItem>
          <MenuItem
            onClick={(): void => {
              setOpen(true);
            }}
          >
            <FormattedMessage id="polygon-selection.box_coorinate-menu_option.text"/>
          </MenuItem>
          <MenuItem
            onClick={(): void => {
              onReset();
              handleClose();
            }}
          >
            <FormattedMessage id="polygon-selection.clear-menu_option.text"/>
          </MenuItem>
        </Menu>
      </Box>
    );
  }
};
