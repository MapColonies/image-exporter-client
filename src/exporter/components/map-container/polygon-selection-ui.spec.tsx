import React from 'react';
import { shallow } from 'enzyme';
import { DrawType } from '@map-colonies/react-components';
import { MenuItem, Menu, Button } from '@map-colonies/react-core';
// eslint-disable-next-line
import '../../../__mocks__/confEnvShim';
import { PolygonSelectionUi } from './polygon-selection-ui';
import { DialogBBox } from './dialog-bbox';

// Enzyme doesnâ€™t work properly with hooks in general, especially for `shallow` so this is the way to mock `react-intl` module.
// Enspired by https://github.com/formatjs/formatjs/issues/1477
jest.mock('react-intl', () => {
  /* eslint-disable */
  const reactIntl = require.requireActual('react-intl');
  const MESSAGES =  require.requireActual('../../../common/i18n');
  const intl = reactIntl.createIntl({
    locale: 'en',
    messages: MESSAGES.default['en'],
  });

  return {
    ...reactIntl,
    useIntl: () => intl,
  };
  /* eslint-enable */
});

/* eslint-disable @typescript-eslint/no-empty-function, @typescript-eslint/no-unsafe-member-access */

const startDraw = jest.fn();
const cancelDraw = jest.fn();
const resetDraw = jest.fn();

describe('Polygon Selection component', () => {

  afterEach(() => {
    startDraw.mockClear();
    cancelDraw.mockClear();
    resetDraw.mockClear();
  });

  it('renders correctly', () => {
    const wrapper = shallow(
      <PolygonSelectionUi
        isSelectionEnabled={false}
        onCancelDraw={() => {}}
        onReset={() => {}}
        onStartDraw={() => {}}
        onPolygonUpdate={(polygon) => {}}
      />
    );
  
    expect(wrapper).toMatchSnapshot();
  });
  
  it('opens the menu with drawing options on open menu button click', () => {
    const wrapper = shallow(
      <PolygonSelectionUi
        onStartDraw={startDraw}
        onCancelDraw={cancelDraw}
        onReset={resetDraw}
        isSelectionEnabled={false}
        onPolygonUpdate={(polygon) => {}}
      />
    );
  
    expect(wrapper.find(Menu).prop('open')).toBe(false); 
  
    wrapper.find(Button).simulate('click', { currentTarget: {} });
  
    expect(wrapper.find(Menu).prop('open')).toBe(true);
  });
  
  it('Box drawing menu items call start draw with correct params on click and closes the menu', () => {
    const wrapper = shallow(
      <PolygonSelectionUi
        onStartDraw={startDraw}
        onCancelDraw={cancelDraw}
        onReset={resetDraw}
        isSelectionEnabled={false}
        onPolygonUpdate={(polygon) => {}}
      />
    );
  
    const openMenuButton = wrapper.find(Button);
   
    openMenuButton.simulate('click', { currentTarget: {} });
  
    wrapper
      .findWhere((n) => {
        return n.type() === MenuItem && 
              n.prop('children').props['id'] === 'polygon-selection.box-menu_option.text';
      })
      .simulate('click');
  
    expect(wrapper.find(Menu).prop('open')).toBe(false);
    expect(startDraw).toHaveBeenCalledWith(DrawType.BOX);
    expect(startDraw).toHaveBeenCalledTimes(1);
  });
  
  it('Custom defined Bbox drawing menu items opens BBox definition dialog on click and closes the menu', () => {
    const wrapper = shallow(
      <PolygonSelectionUi
        onStartDraw={startDraw}
        onCancelDraw={cancelDraw}
        onReset={resetDraw}
        isSelectionEnabled={false}
        onPolygonUpdate={(polygon) => {}}
      />
    );
  
    const openMenuButton = wrapper.find(Button);
    openMenuButton.simulate('click', { currentTarget: {} });
  
    wrapper
      .findWhere((n) => {
        return n.type() === MenuItem && 
              n.prop('children').props['id'] === 'polygon-selection.box_coorinate-menu_option.text';
      })
      .simulate('click');
  
    // expect(wrapper.find(Menu).prop('open')).toBe(false);
    expect(wrapper.find(DialogBBox).prop('isOpen')).toBe(true);
  });
  
  
  it('clicking the clear menu item calls onreset and closes the menu', () => {
    const wrapper = shallow(
      <PolygonSelectionUi
        onStartDraw={startDraw}
        onCancelDraw={cancelDraw}
        onReset={resetDraw}
        isSelectionEnabled={false}
        onPolygonUpdate={(polygon) => {}}
      />
    );
  
    const openMenuButton = wrapper.find(Button);
    openMenuButton.simulate('click', { currentTarget: {} });
  
    wrapper
      .findWhere((n) => {
        return n.type() === MenuItem && 
              n.prop('children').props['id'] === 'polygon-selection.clear-menu_option.text';
      })
      .simulate('click');
  
    expect(wrapper.find(Menu).prop('open')).toBe(false);
    expect(resetDraw).toHaveBeenCalledTimes(1);
  });
  
  it('Cancel draw is shown when IsSelectionEnabled is true, and clicking on the button calls onCancelDraw', () => {
    const wrapper = shallow(
      <PolygonSelectionUi
        onStartDraw={startDraw}
        onCancelDraw={cancelDraw}
        onReset={resetDraw}
        isSelectionEnabled={true}
        onPolygonUpdate={(polygon) => {}}
      />
    );
  
    const button = wrapper
    .findWhere((n) => {
      return n.type() === Button && 
            n.prop('children').props['id'] === 'polygon-selection.cancel-btn.text';
    });
  
    button.simulate('click');
  
    expect(cancelDraw).toHaveBeenCalledTimes(1);
  });
});

