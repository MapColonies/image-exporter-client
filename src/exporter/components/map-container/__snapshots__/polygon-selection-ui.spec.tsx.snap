// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Polygon Selection component renders correctly 1`] = `
<Component
  position="relative"
>
  <Tooltip
    align="bottomLeft"
    content="Draw an Area to export"
  >
    <withRipple(Function)
      className="makeStyles-drawingButton-1"
      onClick={[Function]}
      raised={true}
    >
      <FormattedMessage
        id="polygon-selection.draw-btn.text"
      />
    </withRipple(Function)>
  </Tooltip>
  <DialogBBox
    isOpen={false}
    onPolygonUpdate={[Function]}
    onSetOpen={[Function]}
  />
  <Function
    className="makeStyles-fullWidth-2"
    onClose={[Function]}
    open={false}
  >
    <Function
      onClick={[Function]}
    >
      <FormattedMessage
        id="polygon-selection.box-menu_option.text"
      />
    </Function>
    <Function
      onClick={[Function]}
    >
      <FormattedMessage
        id="polygon-selection.box_coorinate-menu_option.text"
      />
    </Function>
    <Function
      onClick={[Function]}
    >
      <FormattedMessage
        id="polygon-selection.clear-menu_option.text"
      />
    </Function>
  </Function>
</Component>
`;
